### PR TITLE
Allow StorageFactory to wrap encoders and decoders

### PR DESCRIPTION
**Pull request by smarterclayton**
Prepares for allowing encryption at rest of resources as well as any
other lower level optimization we might chose to implement.

Also cleans up a bunch of ugly code.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@enj since you also touched this code recently.  This is to enable secret encryption and put the right hook in at the bottom of the stack.

File: pkg/genericapiserver/server/storage_factory.go [hunk 1]
```diff
@@ -95,6 +107,34 @@ type groupResourceOverrides struct {
 	// of exposing one set of concepts.  autoscaling.HPA and extensions.HPA as a for instance
 	// The order of the slice matters!  It is the priority order of lookup for finding a storage location
 	cohabitatingResources []schema.GroupResource
+	// encoderChainFn is optional and may wrap the provided encoder prior to being serialized.
+	encoderChainFn func(runtime.Encoder) runtime.Encoder
```
Comment (sttts) [inline pkg/genericapiserver/server/storage_factory.go]: s/ChainFn/Decorator/

Comment (smarterclayton) [inline pkg/genericapiserver/server/storage_factory.go]: Makes sense

**Comment by smarterclayton** [comment]
Any other comments?

File: pkg/genericapiserver/server/storage_factory.go [hunk 2]
```diff
@@ -74,7 +74,19 @@ type DefaultStorageFactory struct {
 	APIResourceConfigSource APIResourceConfigSource
 
 	// newStorageCodecFn exists to be overwritten for unit testing.
-	newStorageCodecFn func(storageMediaType string, ns runtime.StorageSerializer, storageVersion, memoryVersion schema.GroupVersion, config storagebackend.Config) (codec runtime.Codec, err error)
+	newStorageCodecFn func(opts StorageCodecOptions) (codec runtime.Codec, err error)
+}
+
+// StorageCodecOptions are the arguments passed to newStorageCodecFn
+type StorageCodecOptions struct {
```
Comment (sttts) [inline pkg/genericapiserver/server/storage_factory.go]: This should be called `StorageCodecConfig`, compare the other structs in `config.go`.

Comment (sttts) [inline pkg/genericapiserver/server/storage_factory.go]: We have the `Options` concept as well inside the `options` subfolder.

File: pkg/genericapiserver/server/storage_factory.go [hunk 3]
```diff
@@ -184,63 +233,37 @@ func (s *DefaultStorageFactory) getStorageGroupResource(groupResource schema.Gro
...
-	}
-
-	etcdMediaType := s.DefaultMediaType
-	if len(groupOverride.mediaType) != 0 {
-		etcdMediaType = groupOverride.mediaType
-	}
-	if len(exactResourceOverride.mediaType) != 0 {
-		etcdMediaType = exactResourceOverride.mediaType
+	// operate on copy
+	config := s.StorageConfig
```
Comment (sttts) [inline pkg/genericapiserver/server/storage_factory.go]: storageConfig
codecConfig

**Comment by sttts** [comment]
For naming consistency in genericapiserver: `StorageCodec{Options -> Config}`

Otherwise, lgtm

File: pkg/genericapiserver/server/storage_factory.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/genericapiserver/server/storage_factory.go]: Ok

**Comment by smarterclayton** [comment]
updated, applying label.

**Comment by enj** [comment]
Seems reasonable.  I am assuming sometime in the future we will add a serializer that does encryption and decryption of specific objects?

**Comment by smarterclayton** [comment]
There are other issues afloat, but other things that might happen here would be an interceptor for fixes to apiVersion on disk, or potentially fixing issues as they come out of the objects

**Comment by smarterclayton** [comment]
@k8s-bot gce etcd3 e2e test this

**Pull request merged**